pkg/util/cluster/cluster.go

```diff
@@ -356,22 +357,103 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
+		return false
+	}
+
+	for _, snet := range ipRanges {
+		if snet.Contains(cidrNet.IP) || cidrNet.Contains(snet.IP) {
+			return true
+		}
+	}
+	return false
+}
+
+// GetIpRangesFromSubnet converts a given azure subnet to a list if IPNets.
+// Because an az subnet can cover multiple ipranges, we need to return a slice
+// instead of just a single ip range. This function never errors. If something
+// goes wrong, it instead returns an empty list.
+func GetIpRangesFromSubnet(subnet mgmtnetwork.Subnet) []*net.IPNet {
+	ipRanges := []*net.IPNet{}
+	if subnet.AddressPrefix != nil {
+		_, ipRange, err := net.ParseCIDR(*subnet.AddressPrefix)
+		if err == nil {
+			ipRanges = append(ipRanges, ipRange)
+		}
+	}
+
+	if subnet.AddressPrefixes == nil {
+		return ipRanges
+	}
+
+	for _, snetPrefix := range *subnet.AddressPrefixes {
+		_, ipRange, err := net.ParseCIDR(snetPrefix)
+		if err == nil {
+			ipRanges = append(ipRanges, ipRange)
+		}
+	}
+	return ipRanges
+}
+
+
+// getAllDevSubnets queries azure to retrieve all subnets assigned the vnet
+// `dev-vnet` in the current resource group
+func (c *Cluster) getAllDevSubnets() ([]mgmtnetwork.Subnet, error) {
+	allSubnets := []mgmtnetwork.Subnet{}
+	availSnetResults, err := c.subnets.List(context.TODO(), c.env.ResourceGroup(), "dev-vnet")
+	if err != nil {
+		return allSubnets, err
+	}
+	allSubnets = append(allSubnets, availSnetResults.Values()...)
+	for availSnetResults.NotDone() {
+		err = availSnetResults.NextWithContext(context.TODO())
+		if err != nil {
+			break
+		}
+		allSubnets = append(allSubnets, availSnetResults.Values()...)
+	}
+
+	return allSubnets, nil
+}
+
+
 func (c *Cluster) generateSubnets() (vnetPrefix string, masterSubnet string, workerSubnet string) {
 	// pick a random 23 in range [10.3.0.0, 10.127.255.0]
 	// 10.0.0.0/16 is used by dev-vnet to host CI
 	// 10.1.0.0/24 is used by rp-vnet to host Proxy VM
 	// 10.2.0.0/24 is used by dev-vpn-vnet to host VirtualNetworkGateway
-	var x, y int
-	// Local Dev clusters are limited to /16 dev-vnet
-	if !c.ci {
-		x, y = 0, 2*rand.Intn(128)
-	} else {
-		x, y = rand.Intn((124))+3, 2*rand.Intn(128)
+
+	allSubnets, err := c.getAllDevSubnets()
+	if err != nil {
+		c.log.Warnf("Error getting existing subnets. Continuing regardless: %v", err)
+	}
+
+	ipRanges := []*net.IPNet{}
+	for _, snet := range allSubnets {
+		ipRanges = append(ipRanges, GetIpRangesFromSubnet(snet)...)
+	}
+
+	for i := 1; i < 100; i++ {
+		var x, y int
+		// Local Dev clusters are limited to /16 dev-vnet
+		if !c.ci {
+			x, y = 0, 2*rand.Intn(128)
+		} else {
+			x, y = rand.Intn((124))+3, 2*rand.Intn(128)
+		}
+		c.log.Infof("Generate Subnet try: %d\n", i)
+		vnetPrefix = fmt.Sprintf("10.%d.%d.0/23", x, y)
+		masterSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y)
+		workerSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y+1)
+		if !ipRangesContainCIDR(ipRanges, workerSubnet) && !ipRangesContainCIDR(ipRanges, masterSubnet) {
+			return vnetPrefix, masterSubnet, workerSubnet
+		}
```

Comment:
Taking into account the consideration I mentioned about the _ipRangesContainCIDR_, this could be changed to:

```go
workerSubnetOverlaps, err := ipRangesContainCIDR(ipRanges, workerSubnet)
if err != nil || workerSubnetOverlaps {
    continue
}
masterSubnetSubnetOverlaps, err := ipRangesContainCIDR(ipRanges, masterSubnet)
if err != nil || masterSubnetSubnetOverlaps {
    continue
}

return vnetPrefix, masterSubnet, workerSubnet
```
which is a bit more explicit and can not produce the false positive when there is an error in _ipRangesContainCIDR_

🙂

Reply:
We could go even one step further and return an error from _generateSubnets()_ when after trying 100 times it was not possible to get valid results. 

Right now we are returning the last values of vnetPrefix, masterSubnet, workerSubnet that we retrieved. It is very unlikely that this happens but we could be returning the wrong values without explicitly stating that. If we want to overcome that, we can simply return an error value outside the loop and check the error from the outside, like:
```go
func (c *Cluster) generateSubnets() (vnetPrefix string, masterSubnet string, workerSubnet string, err error) {
	allSubnets, err := c.getAllDevSubnets()
	if err != nil {
		c.log.Warnf("Error getting existing subnets. Continuing regardless: %v", err)
	}

	ipRanges := []*net.IPNet{}
	for _, snet := range allSubnets {
		ipRanges = append(ipRanges, GetIpRangesFromSubnet(snet)...)
	}

	nTries := 100
	for i := 1; i < nTries; i++ {
		var x, y int
		if !c.ci {
			x, y = 0, 2*rand.Intn(128)
		} else {
			x, y = rand.Intn(124)+3, 2*rand.Intn(128)
		}
		c.log.Infof("Generate Subnet try: %d\n", i)
		vnetPrefix = fmt.Sprintf("10.%d.%d.0/23", x, y)
		masterSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y)
		workerSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y+1)

		workerSubnetOverlaps, err := ipRangesContainCIDR(ipRanges, workerSubnet)
		if err != nil || workerSubnetOverlaps {
			continue
		}

		masterSubnetSubnetOverlaps, err := ipRangesContainCIDR(ipRanges, masterSubnet)
		if err != nil || masterSubnetSubnetOverlaps {
			continue
		}

		return vnetPrefix, masterSubnet, workerSubnet, nil
	}

	return "", "", "", fmt.Errorf("was not able to generate master and worker subnets after %v tries", nTries)
}
```

This has the benefit that clearly returns the zeroed default values of the strings, which  is good practice in Go when returning an error, instead of the last value retrieved. 
Maybe not needed in this case but just an idea.
🙂

Reply:
+1 to adding `error` to the return values of `generateSubnets`